src/utils/spark.py

```diff
@@ -220,3 +229,59 @@ def repartition(self) -> None:
             file_options=file_options,
             max_rows_per_file=5 * 10**6,
         )
+
+    def upload(self) -> None:
+        """
+        Upload the final partitioned Parquet files to S3. This clears the
+        remote S3 equivalent of each local partition prior to upload in order
+        to prevent orphan files. It also clears the local directory for the
+        table on completion.
+        """
+        table_dir = Path(self.final_dir)
+        s3_root_prefix = Path(self.session.s3_prefix)
+        s3_table_prefix = s3_root_prefix / strip_table_prefix(self.table_name)
+
+        # List all files and directories in the local table output directory
+        # Example table_files: { "taxyr=2020/part-0.parquet" }
+        # Example table_subdirs: { "taxyr=2020/" }
+        table_files = set()
+        table_subdirs = set()
+        for file in table_dir.rglob("*.parquet"):
+            if file.is_file():
+                file_key = file.relative_to(table_dir)
+                dir_key = file.relative_to(table_dir).parent
+                table_files.add(file_key)
+                table_subdirs.add(dir_key)
+
+        # For each subdirectory in the local output, purge the equivalent
+        # subdirectory in S3. This is to prevent stale files with different S3
+        # keys from hanging around and polluting our data in Athena
+        paginator = self.session.s3_client.get_paginator("list_objects_v2")
+        s3_files_to_delete = set()
+        for subdir in table_subdirs:
+            for page in paginator.paginate(
+                Bucket=self.session.s3_bucket,
+                Prefix=(s3_table_prefix / subdir).as_posix(),
+            ):
+                for obj in page.get("Contents", []):
+                    s3_files_to_delete.add(obj["Key"])
+
+        if s3_files_to_delete:
+            delete_objects = [{"Key": f"{key}"} for key in s3_files_to_delete]
+            self.session.s3_client.delete_objects(
+                Bucket=self.session.s3_bucket,
+                Delete={"Objects": delete_objects},
+            )
```

Comment:
**[Thought, non-blocking]** It strikes me that if we delete the existing data and then run into a problem during the upload to replace it, we might end up in a situation where we wake up in the morning and find there's no data in the bucket. Is this a problem? We could be more defensive by uploading to a different prefix, clearing the old prefix, and then moving the fresh upload to the old prefix, but I'm not sure yet if the added complexity is worth it -- I'll leave it up to you to decide.

Reply:
Good thinking! In bb103cc I used a set diff such that only files not in the local dir are deleted from the remote. That way we delete orphan partitions/files, then overwrite the rest of the files via upload.